### ISSUE TITLE
regression: `RefCell<LineWriter<std::io::stdio::StdoutRaw>>` cannot be shared between threads safely

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#127340]
The error seems probably correct, but it's not clear why this would have started failing now.

* https://crater-reports.s3.amazonaws.com/beta-1.80-4/beta-2024-06-22/reg/clap-io-0.1.0/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.80-4/beta-2024-06-22/reg/fakelogs-0.1.10-75501d4/log.txt

```
[INFO] [stdout] error[E0277]: `RefCell<LineWriter<std::io::stdio::StdoutRaw>>` cannot be shared between threads safely
[INFO] [stdout]    --> src/lib.rs:205:20
[INFO] [stdout]     |
[INFO] [stdout] 205 |                 Ok(Box::new(stdout))
[INFO] [stdout]     |                    ^^^^^^^^^^^^^^^^ `RefCell<LineWriter<std::io::stdio::StdoutRaw>>` cannot be shared between threads safely
[INFO] [stdout]     |
[INFO] [stdout]     = help: the trait `Sync` is not implemented for `RefCell<LineWriter<std::io::stdio::StdoutRaw>>`, which is required by `StdoutLock<'_>: Sync`
[INFO] [stdout]     = note: if you want to do aliasing and mutation between multiple threads, use `std::sync::RwLock` instead
[INFO] [stdout]     = note: required for `ReentrantLockGuard<'_, RefCell<LineWriter<std::io::stdio::StdoutRaw>>>` to implement `Sync`
[INFO] [stdout] note: required because it appears within the type `StdoutLock<'_>`
[INFO] [stdout]    --> /rustc/64a1fe67112931359c7c9a222f08fd206255c2b5/library/std/src/io/stdio.rs:614:12
[INFO] [stdout]     = note: required for the cast from `Box<StdoutLock<'_>>` to `Box<dyn std::io::Write + Sync>`
```

**Comment by workingjubilee** [comment]
Minimized:

```rust
use std::io::{self, Write};

pub fn open_stdout() -> io::Result<Box<dyn Write + Sync + 'static>> {
    let stdout = Box::leak(Box::new(io::stdout()));
    let stdout = stdout.lock();
    Ok(Box::new(stdout))
}
```

**Comment by workingjubilee** [comment]
cargo-bisect-rustc informs us it is https://github.com/rust-lang-ci/rust/commit/fd225fc8ab992fbb00eb7bb188b3333bfb57dcc0

This is collateral damage of https://github.com/rust-lang/rust/issues/121440

...apparently, std said it can have a little bit of unsoundness, as a treat?

cc @joboet @programmerjake

**Comment by workingjubilee** [comment]
We have the options of
- accepting this regression
- making `stdout()` a custom internal reentrant lock type (and hoping we get it right)
- adjusting ReentrantLock until the original impl was correct
- slapping a Mutex into this bad boy

@rustbot label: I-libs-nominated

**Comment by workingjubilee** [comment]
The other one minimizes to this:
```rust
use std::io;

fn chunk() -> Result<(), io::Error> {
    let stdout = io::stdout();
    let mut handle = stdout.lock();
    write_csv(&mut handle)
}

pub fn write_csv(_w: &mut (dyn io::Write + Sync)) -> Result<(), io::Error> {
    Ok(())
}

```

In both cases, we can see that the problem is _only_ caused if someone deliberately erases the type of the StdoutLock and instead makes it a `&mut dyn io::Write + Sync`, a transition it can no longer fulfill.

**Comment by Ciel-MC** [comment]
Can we try and see how much of crates break if we do 1? 2 feels like too much effort for little gain and 3 is not very pretty.

Disclaimer: Absolute stdlib noob, just offering my first thoughts because it’s brought into my attention

**Comment by workingjubilee** [comment]
It seems only 2 crates with no dependents on crates.io, one of which seems to be an abandoned toy project, which _strongly_ suggests people generally aren't doing this. There's lots to recommend taking the "shrug and do nothing" option, I just wanted to be thorough.

If 3 ("fix the Guard to be actually-Sync") is doable without sacrificing large parts of ReentrantLock's functionality (or performance?) I would say it's the hands-down winner. I just have given it literally zero thought, though it sounds more like wishful thinking than plausible.

**Comment by joboet** [comment]
Fixing `ReentrantLockGuard` is pretty much out of the question. I'd say that the most common type to put in a `ReentrantLock` is a `RefCell`, which is, as we can see here, not `Sync`.

Fixing `StdoutLock` doesn't really make sense, as it obviates the point of locking: if we can get a `&StdoutLock` on multiple threads, we have to make the inner state a `Mutex` (this won't cause any deadlocks by the way, as the `Mutex` is only locked while executing `std`-internal code).

Therefore, I strongly prefer option 1: this was a soundness bug in `std`, so a breaking change is entirely justified.

**Comment by joshtriplett** [comment]
We should relnote this as a soundness fix.

**Comment by workingjubilee** [comment]
Wait, did T-libs ever actually discuss this?

**Comment by workingjubilee** [comment]
Ah, [yes they did on the July 10 meeting](https://hackmd.io/CSacYcDYTQeG37RgDSg1BA#Nominated-rusttf127340-regression-RefCellltLineWriterltstdiostdioStdoutRawgtgt-cannot-be-shared-between-threads-safely), people were like "eh, acceptable".

**Comment by workingjubilee** [comment]
This problem applies to StderrLock too, it's just no one noticed:

![StderrLock in 1.79.0-stable is Sync](https://github.com/user-attachments/assets/21dd0aaa-0dc4-450e-8f18-a384a1f7c967)

![StderrLock in 1.80.0-beta.6 is !Sync](https://github.com/user-attachments/assets/3daeb669-3c62-402c-bf98-bf0fcbafb885)

**Comment by Mark-Simulacrum** [comment]
This was mentioned in the release notes (currently last line of compat notes):

> [`std::sync::ReentrantLockGuard<T>` is no longer `Sync` if `T: !Sync`](https://github.com/rust-lang/rust/pull/125527) which means [`std::io::StdoutLock` and `std::io::StderrLock` are no longer Sync](https://github.com/rust-lang/rust/issues/127340)

Closing.

**Comment by workingjubilee** [comment]
Fixed in clap-io 0.1.1
